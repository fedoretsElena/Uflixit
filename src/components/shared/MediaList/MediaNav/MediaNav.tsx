import React from 'react';
import { FaStepBackward, FaStepForward } from 'react-icons/fa';

interface IProps {
    curr: number,
    length: number,
    uploadPage: Function
}

function generateArr(length: number): number[] {
    const arr: number[] = [];

    for (let i = 1; i <= length; i++) {
        arr.push(i);
    }

    return arr;
}

const MediaNav = ({length, curr, uploadPage}: IProps) => (
    <div className="btn-group btn-group-sm mr-2 pr-1">
        {curr - 1 >=1 && <button type="button"
                key='back'
                className='btn btn-secondary mr-1'
                onClick={() => uploadPage(--curr)}
        >
            <FaStepBackward />
        </button>
        }

        {generateArr(length).map((step: number) => (
            <button type="button"
                    key={step}
                    className={`btn ${ step === curr ? 'btn-primary' : 'btn-secondary' }`}
                    onClick={() => step !== curr ? uploadPage(step) : console.log('This page is already opened')}
            >
                { step }
            </button>
        ))}

        {curr + 1 <= length && <button type="button"
                                key='forward'
                                className='btn btn-secondary ml-1'
                                onClick={() => uploadPage(++curr)}
        >
            <FaStepForward/>
        </button>
        }
    </div>
);

export default MediaNav;