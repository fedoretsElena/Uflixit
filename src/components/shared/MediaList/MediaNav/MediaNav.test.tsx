import MediaNav, { checkCurrPage, generateArr } from "./MediaNav";

describe("MediaNax component", () => {

    it("should render correctly", () => {
        const component = shallow(
            <MediaNav curr={1} length={2} uploadPage={(arg) => arg}/>
        );

        expect(component).toMatchSnapshot();
    });

    it("should format template correctly", () => {
        const component = mount(
            <MediaNav curr={1} length={2} uploadPage={(arg) => arg}/>
        );

        const prevBtn = component.find("button").first().text();

        expect(prevBtn).toEqual("1");
    });

    describe("clicks on nav btns", () => {
        let component;
        let spy;

        beforeEach(() => {
            spy = sinon.spy();
            component = mount(
                <MediaNav curr={2} length={4} uploadPage={spy} />
            );
        });

        it("uploadPage should be called after click on next button", () => {
            // const spy = jest.spyOn(args, "uploadPage"); <= it's working too

            component
                .find("button")
                .last()
                .simulate("click");

            expect(spy.calledWith(3)).toBe(true);
        });

        it("uploadPage should be called after click on prev button", () => {

            component
                .find("button")
                .first()
                .simulate("click");

            expect(spy.calledWith(1)).toBe(true);
        });

        it("should call checkCurrPage after click on navigation items", () => {

            component
                .find("button")
                .at(1)
                .simulate("click");

            expect(spy.calledOnce).toBe(true);
        });
    });
});

describe("generateArr function", () => {

    it("should return navigation arr", () => {
        expect(generateArr(5, 3))
            .toEqual(expect.arrayContaining([2, 3, 4]));
    });

    it("should return length = 3", () => {
        expect(generateArr(5, 3)).toHaveLength(3);
    });
});

describe("checkCurrPage function", () => {
    let mockArg: { curr: number, next: number, callback: () => void };
    let spyOnCallback: any;

    beforeEach(() => {
        mockArg = {
            curr: 1,
            next: 2,
            callback: () => {
            }
        };

        spyOnCallback = jest.spyOn(mockArg, "callback");
    });

    afterEach(() => {
        spyOnCallback = null;
    });

    it("should run callback if current page isn't equal next", () => {
        checkCurrPage(mockArg.curr, mockArg.next, () => mockArg.callback());
        expect(spyOnCallback).toHaveBeenCalled();
    });

    it("shouldn't run callback if current page equal next", () => {
        mockArg.curr = 2;

        checkCurrPage(mockArg.curr, mockArg.next, () => mockArg.callback());

        expect(spyOnCallback).not.toHaveBeenCalled();
    });
});