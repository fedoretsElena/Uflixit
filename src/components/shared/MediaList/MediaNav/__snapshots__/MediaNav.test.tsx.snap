// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaNax component should render correctly 1`] = `
<div
  className="btn-group btn-group-sm mr-2 pr-1 ml-auto"
>
  <button
    className="btn btn-primary"
    key="1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="btn btn-secondary"
    key="2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="btn btn-secondary ml-1"
    key="forward"
    onClick={[Function]}
    type="button"
  >
    <FaStepForward />
  </button>
</div>
`;
